crates/core/src/error/summary.rs

```diff
@@ -0,0 +1,617 @@
+
+    /// Collection of non-critical warnings   
+    issues: Issues,
+
+    /// Optional custom metrics collected during execution
+    metrics: Metrics,
+
+    /// Display this data
+    display: HashSet<DisplayOptionKind>,
+
+    /// Log enabled
+    log_enabled: bool,
+}
+
+impl Summary {
+    /// Constructor to create an initial empty Summary
+    pub fn new(context: impl Into<EcoString>) -> Self {
+        Self {
+            context: context.into(),
+            start_time: Instant::now(),
+            end_time: None,
+            issues: Issues::default(),
+            metrics: BTreeMap::default(),
+            display: HashSet::from([DisplayOptionKind::default()]),
+            log_enabled: false,
+        }
+    }
+
+    /// Marks the summary as completed, capturing the end time.
+    pub fn complete(&mut self) {
+        self.end_time = Some(Instant::now());
+    }
+
+    pub fn enable_log(&mut self) -> Self {
+        self.log_enabled = true;
+        self.clone()
+    }
+
+    pub fn contains_error(&self) -> bool {
+        self.issues.contains_key(&IssueCategory::Error)
+            && !self.issues[&IssueCategory::Error].is_empty()
+    }
+
+    pub fn add_error(
+        &mut self,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        self.add_issue(IssueCategory::Error, scope, message, root_cause);
+    }
+
+    pub fn add_warning(
+        &mut self,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        self.add_issue(IssueCategory::Warning, scope, message, root_cause);
+    }
+
+    pub fn add_info(
+        &mut self,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        self.add_issue(IssueCategory::Info, scope, message, root_cause);
+    }
+
+    /// Adds a new issue to the summary, condensing similar issues
+    pub fn add_issue(
+        &mut self,
+        category: IssueCategory,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        let root_cause = root_cause.into();
+        let message = message.into();
+
+        if self.log_enabled {
+            // We ignore the result here, as we don't want to propagate the error
+            _ = Self::log_issue(scope, category, &message, &root_cause);
+        }
+
+        _ = self
+            .issues
+            .entry(category)
+            .or_default()
+            .entry(scope)
+            .and_modify(|val| {
+                val.count += 1;
+                if val.root_cause.is_none() {
+                    val.root_cause.clone_from(&root_cause);
+                }
+            })
+            .or_insert(CondensedIssue {
+                category,
+                message,
+                count: 1,
+                root_cause,
+            });
```

Comment:
As I see it, when we get in the `and_modify` case, we loose the `message`.
Is that voluntary?

Same we might have != `root_cause` but we'll keep only one.

Wouldn't this lead to "wrong" reporting of some error A that happened N times, but in reality, A happened N1 times and B, C, D, E had the same scope and happened but their "message" and "root_cause" are lost.

I understand we want to aggregate and get a `high level overview`. But as I understand this, we could become `misleading` there.